src/config/config.h

```diff
@@ -0,0 +1,214 @@
+#pragma once
+
+#include "cris/core/utils/defs.h"
+#include "cris/core/utils/logging.h"
+
+#include "config_parser_impl.h"
+
+#include <simdjson.h>
+
+#include <atomic>
+#include <concepts>
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <numeric>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+namespace cris::core {
+
+// To read configuration from file.
+//
+// ConfigFile config_file("my_config.json");
+// auto       my_value = config_file.Get<my_value_t>("my_key")->GetValue();
+//
+// or,
+//
+// auto my_value = *config_file.Get<my_value_t>("my_key", default_value)->GetValue();
+
+class ConfigFile;
+
+class ConfigBase {
+   public:
+    ConfigBase() = default;
+
+    ConfigBase(const ConfigBase&) = delete;
+
+    ConfigBase(ConfigBase&&) = delete;
+
+    ConfigBase& operator=(const ConfigBase&) = delete;
+
+    virtual ~ConfigBase() = default;
+
+    virtual std::string GetName() const = 0;
+
+   protected:
+    virtual void InitValue(simdjson::ondemand::value& val) = 0;
+
+    friend class ConfigFile;
+};
+
+template<class data_t>
+class Config : public ConfigBase {
+   public:
+    std::string GetName() const override { return name_; }
+
+    const data_t& GetValue() const { return data_; }
+
+   protected:
+    explicit Config(const std::string& name);
+
+    void InitValue(simdjson::ondemand::value& val) override;
+
+    friend class ConfigFile;
+
+    const std::string name_;
+    std::atomic<bool> initialized_{false};
+    data_t            data_{};
+};
+
+class ConfigFile {
+   public:
+    explicit ConfigFile(const std::string& filepath);
+
+    template<class data_t>
+    std::shared_ptr<Config<data_t>> Get(const std::string& config_name);
+
+    template<class data_t>
+    std::shared_ptr<Config<data_t>> Get(const std::string& config_name, data_t&& default_value);
+
+   private:
```

Comment:
protected? if you don't have intention to block the use in derived class

Reply:
I don't know what derived class it may have. I don't think there should be one. If it is the case in the future then we will probably modify this class anyways and we might switch to protected if needed.

Reply:
> I don't know what derived class it may have.

If the assumption is not as strong as `final`, better not relying on it.

> I don't think there should be one.

That's a strong assumption, why?

> If it is the case in the future then we will probably modify this class anyways and we might switch to protected if needed.

Code design should justify itself, not relying on downstream.
Here you assume you (or "we", a.k.a. 1st party) are both the lib author and user.
That's not the case in general, and not a good way of writing lib.

Reply:
I think this should probably be a final class, but I am not very sure about this and I don't know if there will be downstreaming  
 (and how) in the future. I don't know if downstreams will need to access those members either. Basically it is all unknown. You can make up any justification but they may turn false somehow. So I would say, for now I don't any reason to make them protected instead of private, so let's keep it that way. I don't know if anyone is going to derive this class either so let's not make it final. If any of these assumptions changed we can then change them (with more educated reasons), or anyone can submit a pull-request/issue if they find necessary (I doubt if there is any, though)

Reply:
(This is not something blocking, I'm fine with private, below is just design discussion.)

Actually I rarely find private without protected to be useful, unless there's also final.
It adds in constraints extra benefit then protected.
For unknown part, I would leave that open for user to decide.
User should have full control on what they get and author only hints about what's API and what's internal with public/protected.

Reply:
Ack.